driverapi/driverapi.go

```diff
@@ -42,6 +42,14 @@ type Driver interface {
 	// Leave method is invoked when a Sandbox detaches from an endpoint.
 	Leave(nid, eid string) error
 
+	// ProgramExternalConnectivity invokes the driver method which does the necessary
+	// programming to allow the external connectivity dictated by the passed options
+	ProgramExternalConnectivity(nid, eid string, options map[string]interface{}) error
```

Comment:
Please note we are not forcing existing remote drivers to implement the new APIs.
Libnetwork will still send the exposed port and mapping information to the network driver as it does today. We decided to go toward a deprecation path for this.

One reason for the new APIs was that in case of container joining/leaving multiple networks which results in a new highest priority default gw, a clean switch-over of port-mapping can only be done decoupling the external connectivity programming from the create/join endpoint.
Otherwise the only available option was to perform a fictitious join/leave probe before the real join, in order to get the info about the default gw.

Then there is an advantage in decoupling this functionality, making it modular in that libnetwork can now control the external connectivity of the networks (as of today, there is no guarantee the driver will honor the `internal` flag)